helix-core/src/main/java/org/apache/helix/model/ResourceConfig.java

```diff
@@ -842,5 +841,89 @@ public ResourceConfig build() {
+              idealState.getMaxPartitionsPerInstance());
+    }
+    if (null == mergedZNRecord
+        .getSimpleField(ResourceConfig.ResourceConfigProperty.NUM_PARTITIONS.name())) {
+      mergedZNRecord
+          .setIntField(ResourceConfigProperty.NUM_PARTITIONS.name(), idealState.getNumPartitions());
+    }
+    if (null == mergedZNRecord
+        .getSimpleField(ResourceConfig.ResourceConfigProperty.STATE_MODEL_DEF_REF.name())) {
+      mergedZNRecord
+          .setSimpleField(ResourceConfig.ResourceConfigProperty.STATE_MODEL_DEF_REF.name(),
+              idealState.getStateModelDefRef());
+    }
+    if (null == mergedZNRecord
+        .getSimpleField(ResourceConfig.ResourceConfigProperty.STATE_MODEL_FACTORY_NAME.name())) {
+      mergedZNRecord
+          .setSimpleField(ResourceConfig.ResourceConfigProperty.STATE_MODEL_FACTORY_NAME.name(),
+              idealState.getStateModelFactoryName());
+    }
+    if (null == mergedZNRecord
+        .getSimpleField(ResourceConfig.ResourceConfigProperty.REPLICAS.name())) {
+      mergedZNRecord.setSimpleField(ResourceConfig.ResourceConfigProperty.REPLICAS.name(),
+          idealState.getReplicas());
+    }
+    if (null == mergedZNRecord
+        .getSimpleField(ResourceConfig.ResourceConfigProperty.MIN_ACTIVE_REPLICAS.name())) {
+      mergedZNRecord.setIntField(ResourceConfig.ResourceConfigProperty.MIN_ACTIVE_REPLICAS.name(),
+          idealState.getMinActiveReplicas());
+    }
+    if (null == mergedZNRecord
+        .getSimpleField(ResourceConfig.ResourceConfigProperty.HELIX_ENABLED.name())) {
+      mergedZNRecord.setBooleanField(ResourceConfig.ResourceConfigProperty.HELIX_ENABLED.name(),
+          idealState.isEnabled());
+    }
+    if (null == mergedZNRecord
+        .getSimpleField(ResourceConfig.ResourceConfigProperty.RESOURCE_GROUP_NAME.name())) {
+      mergedZNRecord
+          .setSimpleField(ResourceConfig.ResourceConfigProperty.RESOURCE_GROUP_NAME.name(),
+              idealState.getResourceGroupName());
+    }
+    if (null == mergedZNRecord
+        .getSimpleField(ResourceConfig.ResourceConfigProperty.RESOURCE_TYPE.name())) {
+      mergedZNRecord.setSimpleField(ResourceConfig.ResourceConfigProperty.RESOURCE_TYPE.name(),
+          idealState.getResourceType());
+    }
+    if (null == mergedZNRecord
+        .getSimpleField(ResourceConfig.ResourceConfigProperty.EXTERNAL_VIEW_DISABLED.name())) {
+      mergedZNRecord
+          .setBooleanField(ResourceConfig.ResourceConfigProperty.EXTERNAL_VIEW_DISABLED.name(),
+              idealState.isExternalViewDisabled());
+    }
+    if (null == mergedZNRecord
+        .getSimpleField(ResourceConfig.ResourceConfigProperty.DELAY_REBALANCE_ENABLED.name())) {
+      mergedZNRecord
+          .setBooleanField(ResourceConfig.ResourceConfigProperty.DELAY_REBALANCE_ENABLED.name(),
+              idealState.isDelayRebalanceEnabled());
+    }
```

Comment:
I see a lot of duplicate code here -

could we just create an array of all `ResourceConfig.ResourceConfigProperty`s and just copy the fields?

Reply:
Some names are different in two list, . And the types are different.

Reply:
I tried.
1. Although the field keys are the same, the get logic is different. If I copypaste directly based on the field names, I might change the IS logic. For example, getReplicas(). It would be safer that I call the explicit IdealState get methods.
Unfortunately, this will lead to a similar code. But if we take this as a temporary backward compatible workaround, I think it is acceptable. For the correctness.
2. The 2 lists are not exactly the same. DELAY_REBALANCE_DISABLED in IS and DELAY_REBALANCE_ENABLED in the ResourceConfig.

To reduce this complex, I propose to add some setXXXIfAbsent in the ZNRecord class. Will update soon.